docs/our-approach-to-data.md

```diff
@@ -0,0 +1,301 @@
+
+```
+page( '/posts/:siteId', ( context ) => {
+	const siteId = context.params.siteId;
+
+	ReactDOM.render(
+		<PostsData siteId={ siteId }>
+			<PostsList siteId={ siteId } />
+		</PostsData>,
+		document.getElementById( 'primary' )
+	);
+} );
+```
+
+Be mindful of the fact that fetching components are an undesirable necessity that we'd like to remove in the future, and as such treat it as though it may eventually be removed altogether. This is why the `siteId` prop is specified on both `<PostsData />` and `<PostsList />` when rendering.
+
+### Selectors
+
+A selector is simply a convenience function for retrieving data out of the global state tree. Since the global state tree is a plain JavaScript object, there's nothing to stop you from accessing data directly, but you may find that creating a selector will help to reduce repetition, improve the semantic meaning of your code, and avoid mistakes. For example, the following two approaches both serve to retrieve an array of posts for a specific site, though you might find the selector more convenient and readable at a glance:
+
+```js
+// Using a selector
+let posts = getSitePosts( state, siteId );
+
+// Navigating the state tree
+let posts = state.sites.sitePosts[ siteId ].map( ( postId ) => state.posts.items[ postId ] );
+```
+
+You'll note in this example that the entire `state` object is passed to the selector. We've chosen to standardize on always sending the entire state object to any selector as the first parameter. This consistency should alleviate uncertainty in calling selectors, as you can always assume that it'll have a similar argument signature. More importantly, it's not uncommon for selectors to need to traverse different parts of the global state, as in the example above where we pull from both the `sites` and `posts` top-level state keys.
+
+What are a few common use-cases for selectors?
+
+- Resolving references: A [normalized state tree](#data-normalization) is ideal from the standpoint of minimizing redundancy and syncronization concerns, but is not as developer-friendly to use. Selectors can be helpful in restoring convenient access to useful objects.
+- Filtering data: You can use a selector to return a subset of a state tree value. For example, a `getJetpackSites` selector could return an array of all known sites filtered to only those which are Jetpack-enabled. 
+ - __Side-note:__ In this case, you could achieve a similar effect with a reducer function aggregating an array of Jetpack site IDs. If you were to take this route, you'd probably want a complementary selector anyways. Caching concerns on selectors can be overcome by using memoization techniques (for example, with a library like [`reselect`](https://github.com/rackt/reselect)).
+
+### Derived Data
+
+__WIP__
+
+- Can/should we use [JavaScript object getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) as part of the state tree in creating computed properties?
```

Comment:
>Wouldn't selectors make sense for retrieving computed properties?

I've intentionally left this section as WIP because I'm not really sold one way or the other yet. I also think this potentially leads us to other questions such as: Should we "fix" API responses to a format consistent with our naming guidelines (i.e. `site.ID` -> `site.id`).

The particular question that was posed was: Developers currently have easy access to a site domain by calling `site.domain`. The `domain` is a derived property based on other fields on the object. What recommendation should we have for access this data moving forward?

A few options I see are:

```js
// 1)
// 
// Components should receive the most minimal data necessary, so pass `domain`
// using a selector which calculates the value based on the site the state tree

export default connect( ( state ) => {
	return {
		url: getSite( state, siteId ).URL,
		domain: getSiteDomain( state, siteId )
	};
} )( SiteLink );

// Pro: Component receives minimal props
// Con: Domain is not as accessible as calling to "site.domain"

// 2)
// 
// Components should receive the minimal set of props, so pass domain, but as
// a destructured property of the site object stored in state

export default connect( ( state, ownProps ) => {
	const { URL: url, domain } = getSite( state, ownProps.siteId );
	return { url, domain };
} )( SiteLink );

// Pro: As accessible as calling to "site.domain"
// Con: Either the site state is not normalized (has redundant computed data),
//      or we must define JavaScript getters in the state tree

// 3)
// 
// Pass the entire site object, including derived properties, to the component

export default connect( ( state, ownProps ) => {
	return { site: getSite( state, ownProps.siteId };
} )( SiteLink );

// Pro: Easy to access any necessary values from component itself
// Con: Not passing minimal set of props
// Con: Site not normalized or JavaScript getters in state tree
```